launcher/ui/themes/ThemeManager.cpp

```diff
@@ -0,0 +1,154 @@
+// SPDX-License-Identifier: GPL-3.0-only
+/*
+ *  Prism Launcher
+ *  Copyright (C) 2022 Tayou <tayou@gmx.net>
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, version 3.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+#include "ThemeManager.h"
+
+#include "ui/themes/SystemTheme.h"
+#include "ui/themes/DarkTheme.h"
+#include "ui/themes/BrightTheme.h"
+#include "ui/themes/CustomTheme.h"
+#include <QDir>
+#include <QDirIterator>
+#include <QIcon>
+#include <QApplication>
+
+#include "Application.h"
+
+#ifdef Q_OS_WIN
+#include <windows.h> 
+// this is needed for versionhelpers.h, it is also included in WinDarkmode, but we can't rely on that. 
+// Ultimately this should be included in versionhelpers, but that is outside of the project.
+#include "ui/WinDarkmode.h"
+#include <versionhelpers.h>
+#endif
+
+ThemeManager::ThemeManager(MainWindow* mainWindow) {
+    m_mainWindow = mainWindow;
+    InitializeThemes();
+}
+
+/// @brief Adds the Theme to the list of themes
+/// @param theme The Theme to add
+/// @return Theme ID
+QString ThemeManager::AddTheme(std::unique_ptr<ITheme> theme) {

```

Comment:
```suggestion
QString ThemeManager::AddTheme(std::unique_ptr<ITheme>&& theme) {

```

not sure if i'm getting this right. Since you're doing `std::move` from this function, you won't be able to rely on it being valid after the fact, right? So, we can communicate that using a && thingy.

Reply:
this is the point of unique_ptr :)
you can only move it, so by copy is the right thing to do to mean this function takes ownership

Reply:
but if you only do a copy, you're not signaling to the user of this function that the pointer they passed to it will no longer have the valid object in it :thinking:

Reply:
exactly, otherwise you would use a shared_ptr or return a reference. this is just to mean that this is taking ownership. In the usage for now, the pointer is not used after being passed to AddTheme

Reply:
in any case with a unique_ptr you should have only 1 owning pointer you can have raw pointers as observers but you shouldn't keep them as the unique_ptr could delete the memory. so it depends on what you want to do, but in this case I find a unique_ptr with the AddTheme working great. as shared_ptr should be aboided if possible

Reply:
so, this function is taking ownership of the unique_ptr with or without copying. This means that, regardless of which one we use, it's gonna do the same stuff. The thing is, using && makes the unique_ptr usage similar to any other object you pass to a function that takes ownership of it. Adding the && just makes it more clear that this is going on. Is this right?

Reply:
nope ^^
you can't copy a unique_ptr you can only move it
&& is a rvalue reference, it means it can only bind to rvalues, but it's just a reference so if you do that you can move inside the function or you may not, the unique_ptr is already stating the fact that it can only be moved, and by value insist on the fact that it will be moved

Reply:
also && is adding a layer of pointer, while just passing by value just copy the pointer

Reply:
this screams bad language design, would be much easier to read code if you just had everything that gets moved using the same syntax, even if it technically makes sense to be this way. But I guess it is what it is :woman_shrugging:

Reply:
well move semantic wasn't built-in from the start, it was added in C++11 ^^
I guess its a habit, `std::move` doesn't move it just casts to an rvalue reference, && either, it's just the rvalue reference and it binds to rvalues, the actual move occurs when calling the move constructor or move assignment operator, which are called when passing an rvalue of the same type to them. also move is a big word, often its just "stealing" so copy of pointers and zeroing the source.
you can have a good grasp at this with CppCon videos, they are really great, they have a "back to basics" series: https://www.youtube.com/watch?v=knEaMpytRMA&t=6 :)

Reply:
CppCon is great, i'll have a look, thanks `\ ^-^/`